### PR TITLE
mon/OSDMonitor: do not alter the "created" epoch of a pg

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: http://tracker.ceph.com/issues/19787
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -3205,18 +3205,15 @@ void OSDMonitor::update_creating_pgs()
 		     << " acting_primary:" << last_acting_primary
 		     << " -> " << acting_primary << dendl;
 	    // note epoch if the target of the create message changed.
-	    // creating_pgs is updated here instead of in
-	    // scan_for_creating_pgs() because we don't have the updated pg
-	    // mapping by then.
-	    created = mapping.get_epoch();
+	    mapped = mapping.get_epoch();
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: I think here we also need to remove the pg from it's old position.. pgs_by_epoch.second.erase(pgid)?

Comment (liewegas) [inline src/mon/OSDMonitor.cc]: oh, nevermind, you rebuild teh whole thing!

Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: yeah, assuming the set of pg being created should be smallish in practice.

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -3205,18 +3205,15 @@ void OSDMonitor::update_creating_pgs()
...
           }
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: else {
  mapped = pgs_by_epoch.first;
}

Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: no, it's not necessary. `mapped` is `pg.second.first` from the very beginning, if it's not changed after that.

**Comment by tchaikov** [comment]
@liewegas can we fix this issue first and then think about how to improve the performance or if it's necessary?

the `creating_pgs_by_osd_epoch` is organized in a different way than `pg_creatings`, i am afraid that updating the map instead rebuilding it from scratch would be much more complicated: two passes: 1) to iterate in the map to trim the pgs not exist in creating_pg, 2) to iterate in pg_creatings to add/update pg being created.  or we can walk thru the pools and cross check `creating_pgs_by_osd_epoch` and `creating_pgs` to find stale or non-existent pgs.

i am not sure it is worthwhile introducing the complicity.

File: src/mon/OSDMonitor.cc (hunk 2 above)
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: oh, i was wrong. will fix.

**Comment by liewegas** [comment]
yeah, find to ignore the performance.  it's likely not a big deal.

**Comment by liewegas** [comment]
looks good, except the last (emplace) commit has the hunk fixing the mapped value mixed into it.

**Comment by tchaikov** [comment]
i just tried to update the `pgs_by_osd_epoch` in-place, the size of this function tripled. if we find it's a bottleneck. i will trade the simplicity for the performance.

**Comment by tchaikov** [comment]
@liewegas fixed and repushed.

**Comment by liewegas** [comment]
yeah sounds good!

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/misc.sh:169: jq_success:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/misc.sh:251: TEST_mon_features:  return 1
```

retest this please.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-04-28_19:45:54-rados-wip-sage-testing---basic-smithi/

**Pull request merged**